src/middleware/cache/index.ts

```diff
@@ -118,6 +135,10 @@ export const cache = (options: {
     }
     addHeader(c)
     const res = c.res.clone()
+    if (options.duration) {
+      const expiration = Date.now() + options.duration * 1000
+      res.headers.set(expirationHeader, String(expiration))
```

Comment:
https://github.com/honojs/hono/pull/3318 will solve the issue, right?

Reply:
Thanks for confirming.
I guess #3318 doesn't solve the problem.

If you run the following app with deno, you will see that `TypeError: Headers are immutable.`

```ts
import { Hono } from './src/hono'
import { cache } from './src/middleware/cache'

const app = new Hono()
app.use(cache({
  cacheName: Math.random().toString(),
  cacheControl: 'max-age=3600',
  wait: true,
  duration: 100,
}))
app.get('/', (c) => fetch('https://hono.dev/'))

export default app
```

```
TypeError: Headers are immutable.
    at Headers.set (ext:deno_fetch/20_headers.js:425:13)
    at cache (file:///Users/taku/src/github.com/honojs/hono/src/middleware/cache/index.ts:135:19)
    at eventLoopTick (ext:core/01_core.js:168:7)
    at async dispatch (file:///Users/taku/src/github.com/honojs/hono/src/compose.ts:74:17)
    at async file:///Users/taku/src/github.com/honojs/hono/src/hono-base.ts:444:25
    at async ext:deno_http/00_serve.ts:364:18
```

This is a very redundant cloning process, but to avoid this, the following should be done.

```diff
diff --git a/src/middleware/cache/index.ts b/src/middleware/cache/index.ts
index 6c04207d..6d122589 100644
--- a/src/middleware/cache/index.ts
+++ b/src/middleware/cache/index.ts
@@ -130,7 +132,10 @@ export const cache = (options: {
       return
     }
     addHeader(c)
-    const res = c.res.clone()
+    const res = new Response(c.res.clone().body, {
+      status: c.res.status,
+      headers: c.res.headers,
+    })
     if (options.duration) {
       res.headers.set(cachedTime, String(Date.now()))
     }
```

Alternatively, if one were to avoid this overhead in the normal case, this might be a good approach.

```diff
diff --git a/src/middleware/cache/index.ts b/src/middleware/cache/index.ts
index 6c04207d..0be76724 100644
--- a/src/middleware/cache/index.ts
+++ b/src/middleware/cache/index.ts
@@ -130,9 +132,22 @@ export const cache = (options: {
       return
     }
     addHeader(c)
-    const res = c.res.clone()
+    let res = c.res.clone()
     if (options.duration) {
-      res.headers.set(cachedTime, String(Date.now()))
+      try {
+        res.headers.set(cachedTime, String(Date.now()))
+      } catch (e) {
+        if (e instanceof TypeError && e.message.includes('immutable')) {
+          // `res` is immutable (probably a response from a fetch API), so retry with a new response.
+          res = new Response(res.body, {
+            headers: res.headers,
+            status: res.status,
+          })
+          res.headers.set(cachedTime, String(Date.now()))
+        } else {
+          throw e
+        }
+      }
     }
     if (options.wait) {
       await cache.put(key, res)
```

It may be possible to provide a utility method, but that is up to preference.

Reply:
Thank you for the fix!